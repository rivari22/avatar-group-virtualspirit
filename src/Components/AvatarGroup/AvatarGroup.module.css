.container {
  display: flex;
}
.avatar {
  border-radius: 50%;
  outline: 2px solid white;
  outline-offset: -1px;
  display: flex;
  justify-content: center;
  align-items: center;
}
.empty-image {
  background-color: #E7EAF4;
}

.card-size-xs {
  width: 32px;
  height: 32px;
}
.card-size-sm {
  width: 46px;
  height: 46px;
}
.card-size-md {
  width: 64px;
  height: 64px;
}
.card-size-lg {
  width: 120px;
  height: 120px;
}

.font-card {
  color: #767B87;
  margin-bottom: 4px;
}
.font-card-xs {
  font-size: 14px;
  font-weight: bold;
}
.font-card-sm {
  font-size: 20px;
  font-weight: bold;
}
.font-card-md {
  font-size: 30px;
  font-weight: bold;
}
.font-card-lg {
  font-size: 56px;
  font-weight: bold;
}

.avatar-gap:first-child {
  margin-left: 0px;
}
.avatar-gap-xs:not(first-child) {
  margin-left: -8px;
}
.avatar-gap-sm:not(first-child) {
  margin-left: -12px;
}
.avatar-gap-md:not(first-child) {
  margin-left: -18px;
}
.avatar-gap-lg:not(first-child) {
  margin-left: -34px;
}
