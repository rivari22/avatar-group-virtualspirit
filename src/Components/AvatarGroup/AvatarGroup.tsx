import React from "react";
import classNames from "classnames";

import styles from "./AvatarGroup.module.css";

type SizeType = "xs" | "sm" | "md" | "lg";

type CardImageProps = {
  size?: SizeType;
  image?: string;
  name: string;
};

const CardImage = ({ size, image, name }: CardImageProps) => {
  if (!!image) {
    return (
      <img
        src={image}
        alt="avatar"
        className={classNames(
          styles["avatar-gap"],
          styles["avatar"],
          styles[`card-size-${size}`],
          styles[`avatar-gap-${size}`]
        )}
      />
    );
  }

  return (
    <div
      className={classNames(
        styles["avatar-gap"],
        styles["empty-image"],
        styles["avatar"],
        styles[`card-size-${size}`],
        styles[`avatar-gap-${size}`]
      )}
    >
      <h3
        className={classNames(styles[`font-card-${size}`], styles["font-card"])}
      >
        {name}
      </h3>
    </div>
  );
};

type ItemAvatarGroups = Omit<CardImageProps, "size">;

type AvatarGroupProps = {
  size: SizeType;
  maxLength?: number;
  items: Array<ItemAvatarGroups>;
};

const AvatarGroup = ({ size = "md", maxLength, items }: AvatarGroupProps) => {
  return (
    <div className={styles["container"]}>
      {items
        ?.slice(0, maxLength || -1)
        .map((item: ItemAvatarGroups, index: number) => (
          <CardImage
            key={index}
            image={item.image}
            size={size}
            name={item?.name?.[0] + item?.name?.split(" ")?.[1]?.[0]}
          />
        ))}
      {!!maxLength && items?.length > maxLength && (
        <CardImage size={size} name={`+${items?.length - maxLength}`} />
      )}
    </div>
  );
};

export default AvatarGroup;
