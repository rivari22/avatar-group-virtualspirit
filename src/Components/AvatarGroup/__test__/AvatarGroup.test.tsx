import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";

import AvatarGroup, { AvatarGroupProps } from "../AvatarGroup";
import { dataMock } from "Mocks";

const mockProps: AvatarGroupProps = {
  items: dataMock.items,
  size: "md",
  maxLength: 19,
};

describe("default component should render properly", () => {
  test("renders avatar group", () => {
    render(<AvatarGroup {...mockProps} />);
    const element = screen.getByTestId("container-avatar-group-test");
    expect(element).toBeInTheDocument();
  });
});

describe("each props should be test as expected", () => {
  test("props items image should be there", () => {
    render(<AvatarGroup {...mockProps} />);
    expect(screen.getAllByRole("img")[0]).toHaveAttribute(
      "src",
      dataMock.items[0].image
    );
  });

  test("props size should be md", () => {
    render(<AvatarGroup {...mockProps} />);
    expect(screen.getAllByRole("img")[0]).toHaveClass("card-size-md");
  });

  test("props name should be TB", () => {
    render(<AvatarGroup {...mockProps} />);
    expect(screen.getByRole("heading")).toHaveTextContent("TB");
  });
});

describe("capture a component as snapshot file", () => {
  test("renders snapshot correctly", () => {
    const tree = renderer.create(<AvatarGroup {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
