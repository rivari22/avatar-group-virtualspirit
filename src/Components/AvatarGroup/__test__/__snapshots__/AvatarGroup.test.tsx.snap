// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`capture a component as snapshot file renders snapshot correctly 1`] = `
<div
  className="container"
  data-testid="container-avatar-group-test"
>
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <div
    className="avatar-gap empty-image avatar card-size-md avatar-gap-md"
  >
    <h3
      className="font-card-md font-card"
    >
      TB
    </h3>
  </div>
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
  <img
    alt="avatar"
    className="avatar-gap avatar card-size-md avatar-gap-md"
    src="avatar.png"
  />
</div>
`;
